src/Confluent.Kafka/Producer.cs

```diff
@@ -796,435 +992,243 @@ public TopicPartitionOffsetError TopicPartitionOffsetError
-            => Poll(timeout.TotalMillisecondsAsInt());
+            public string Topic;
 
+            public byte[] Key;
 
-        internal int Flush(int millisecondsTimeout)
-            => kafkaHandle.Flush(millisecondsTimeout);
+            public byte[] Value;
 
+            public Action<DeliveryReportResult> Handler;
 
-        /// <summary>
-        ///     <see cref="Confluent.Kafka.Producer{TKey, TValue}" />
-        /// </summary>
-        internal int Flush(TimeSpan timeout)
-            => kafkaHandle.Flush(timeout.TotalMillisecondsAsInt());
-
-
-        /// <summary>
-        ///     <see cref="Confluent.Kafka.Producer{TKey, TValue}" />
-        /// </summary>
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
-
-
-        /// <summary>
-        ///     <see cref="Confluent.Kafka.Producer{TKey, TValue}" />
-        /// </summary>
-        protected virtual void Dispose(bool disposing)
-        {
-            // Calling Dispose a second or subsequent time should be a no-op.
-            lock (disposeHasBeenCalledLockObj)
-            { 
-                if (disposeHasBeenCalled) { return; }
-                disposeHasBeenCalled = true;
-            }
-
-            if (disposing)
+            public void HandleDeliveryReport(UntypedDeliveryReport deliveryReport)
             {
-                if (!this.manualPoll)
+                if (deliveryReport == null)
                 {
-                    callbackCts.Cancel();
-                    try
-                    {
-                        // Note: It's necessary to wait on callbackTask before disposing kafkaHandle
-                        // since the poll loop makes use of this.
-                        callbackTask.Wait();
-                    }
-                    catch (AggregateException e)
-                    {
-                        if (e.InnerException.GetType() != typeof(TaskCanceledException))
-                        {
-                            throw e.InnerException;
-                        }
-                    }
-                    finally
+                    return;
+                }
+
+                var dr = new DeliveryReportResult
+                {
+                    TopicPartitionOffsetError = deliveryReport.TopicPartitionOffsetError,
+                    Message = new Message
                     {
-                        callbackCts.Dispose();
+                        Key = Key,
+                        Value = Value,
+                        Timestamp = deliveryReport.Message == null 
+                            ? new Timestamp(0, TimestampType.NotAvailable) 
+                            : deliveryReport.Message.Timestamp,
+                        Headers = deliveryReport.Message?.Headers
                     }
-                }
+                };
+                // topic is cached in this object, not set in the deliveryReport to avoid the 
```

Comment:
you're right, I could reuse the generic ones for the non-generic case. I think this will be affected but the up-coming iteration on the API though, so leaving for now.